typescript/sdk/src/ism/EvmIsmModule.hardhat-test.ts

```diff
@@ -0,0 +1,720 @@
+    );
+
+    // get addresses of factories for the chain
+    factoryContracts = contractsMap[chain];
+    factoryAddresses = Object.keys(factoryContracts).reduce((acc, key) => {
+      acc[key] =
+        contractsMap[chain][key as keyof ProxyFactoryFactories].address;
+      return acc;
+    }, {} as Record<string, Address>) as HyperlaneAddresses<ProxyFactoryFactories>;
+
+    // legacy HyperlaneIsmFactory is required to do a core deploy
+    const legacyIsmFactory = new HyperlaneIsmFactory(
+      contractsMap,
+      multiProvider,
+    );
+
+    // mailbox
+    mailboxAddress = (
+      await new TestCoreDeployer(multiProvider, legacyIsmFactory).deployApp()
+    ).getContracts(chain).mailbox.address;
+
+    // new mailbox
+    newMailboxAddress = (
+      await new TestCoreDeployer(multiProvider, legacyIsmFactory).deployApp()
+    ).getContracts(chain).mailbox.address;
+
+    // example routing config
+    exampleRoutingConfig = {
+      type: IsmType.ROUTING,
+      owner: await multiProvider.getSignerAddress(chain),
+      domains: Object.fromEntries(
+        testChains
+          .filter((c) => c !== TestChainName.test4)
+          .map((c) => [c, randomMultisigIsmConfig(3, 5)]),
+      ),
+    };
+  });
+
+  // Wrapper for creating an ISM module
+  // include option for mailbox address override
+  async function createIsmModule({
+    config,
+    mailbox = mailboxAddress,
+  }: {
+    config: IsmConfig;
+    mailbox?: Address;
+  }): Promise<EvmIsmModule> {
+    return EvmIsmModule.create({
+      chain,
+      config,
+      deployer: ismFactoryDeployer,
+      factories: factoryAddresses,
+      mailbox,
+      multiProvider,
+    });
+  }
+
+  // Wrapper for checking whether ISM module matches a given config
+  // include option for mailbox address override
+  async function ismModuleMatchesConfig({
+    ism,
+    config,
+    mailbox = mailboxAddress,
+    configured = true,
+  }: {
+    ism: EvmIsmModule;
+    config: IsmConfig;
+    mailbox?: Address;
+    configured: boolean;
+  }): Promise<boolean> {
+    return moduleMatchesConfig(
+      chain,
+      ism.serialize().deployedIsm,
+      config,
+      multiProvider,
+      factoryContracts,
+      mailbox,
+      configured,
+    );
+  }
+
+  // Wrapper for create a new multiprovider with an impersonated account
+  async function impersonateAccount(account: Address): Promise<MultiProvider> {
+    await hre.ethers.provider.send('hardhat_impersonateAccount', [account]);
+    await fundingAccount.sendTransaction({
+      to: account,
+      value: hre.ethers.utils.parseEther('1.0'),
+    });
+    return MultiProvider.createTestMultiProvider({
+      signer: hre.ethers.provider.getSigner(account),
+    });
+  }
+
+  async function logAndSendTransactions(txs: AnnotatedEthersV5Transaction[]) {
+    for (const tx of txs) {
+      if (tx.annotation) {
+        console.log(tx.annotation);
+      }
+      await multiProvider.sendTransaction(chain, tx.transaction);
+    }
+  }
```

Comment:
if Annotated extended the base tx type, you could move this logging and type into the multi provider in a backward compatible manner

Reply:
good shout

Reply:
had a quick attempt at this but the problem is multiprovider accepts either a PopulatedTx or a Promise<PopulatedTx> and i'm not sure we want to await just to grab an annotation out for logging